### PR TITLE
feat/mercurius federation export #2

### PR DESCRIPTION
**Pull request by codeflyer**
Import current mercurius gateway implementation

#2 

Main part covered

* Created as a plugin, to use in replacemennt of `mercurius` with the same params.
* Test coverage
* Types defined and tested
* 100% compatible with the `mercurius` version

Next step:
* add benchmarks
* open issues to cover TODOs and coverage ignore

### CODE REVIEW

File: README.md
Comment (mcollina) [inline README.md:5]: I think the end goal is to split the implementation of Apollo Federation out from Mercurius.

File: README.md
Comment (mcollina) [inline README.md:14]: What name are we going to use for this modulev

File: README.md
Comment (mcollina) [inline README.md:105]: My goal is to remove the integrated mode altogether.

File: lib/gateway.js
Comment (mcollina) [inline lib/gateway.js:83]: This looks like a code smell, why did you add it?

File: lib/gateway.js
Comment (mcollina) [inline lib/gateway.js:174]: This is fundamentally incorrect. Don't mix promises and callbacks.

File: README.md
Comment (mcollina) [inline README.md:126]: This might be better to be its own module, it seems simple enough.

File: README.md
Comment (mcollina) [inline README.md:151]: I think the gateway API might be better served as a total wrapper for mercurius instead.

File: package.json
Comment (mcollina) [inline package.json:62]: Avoid peerDependencies. They do not work the same across package managers :/.

File: package.json
Comment (mcollina) [inline package.json:68]: why this is a pinned version?

File: package.json
Comment (codeflyer) [reply package.json]: Removed

File: package.json
Comment (codeflyer) [reply package.json]: Fixed

File: README.md
Comment (codeflyer) [reply README.md:126]: Yes, this can be separated easily.

Do you think it's better to deploy 2 packages from this project? Or create a separated one?

File: README.md
Comment (codeflyer) [reply README.md:151]: Do you mean something like this?

```
  app.register(require('mercurius-federation-gateweay'), {
    gateway: {
      services: [
        {
          name: 'user',
          url: 'http://localhost:4001/graphql'
        },
        {
          name: 'post',
          url: 'http://localhost:4002/graphql'
        }
      ]
    },
    ...otherMercuriusProps
  })
  ```

This will be helpful on managing some other stuff. (see comment about `onAppRegister(app.graphql)`)

File: README.md
Comment (codeflyer) [reply README.md:14]: Some proposal

gateway
* `@mercurius/gateway`
* `@mercurius/apollo-gateway`

schemabuilder
* `@mercurius/federation-schema`
* `@mercurius/apollo-federation-schema`

File: README.md
Comment (mcollina) [reply README.md:151]: Exactly.

Comment (mcollina) [reply README.md:151]: Also `@mercurius/gateway`

Comment (codeflyer) [reply README.md:151]: Ok, perfect.

**Review by jonnydgreen — COMMENTED** [review]
Love this idea! :) I have a few generic questions:
 - Do you have plans for TypeScript support (inc. type tests) in subsequent PRs?
 - Do you have plans for benchmarking support in subsequent PRs?
 - Do you have plans for more in depth docs in subsequent PRs that covers things like:
   - Plugin API Options
   - Gateway-specific hooks
   - Behaviour and further examples
 - Did you consider starting to use terminology from this proposal? https://github.com/graphql/composite-schemas-wg/pull/34/files#diff-c42a05dfff7873800ca86ea805b61b7a52aefb8f9462d0e4adb5753a037d4bfaR122 - WDYT?
 - This could also be a good opportunity to identity any existing TODOs/missing code coverage and adding related GH issues for them - WDYT?

**Comment by codeflyer** [comment]
I've reconverted the Pr in Draft, there are changes that will take a bit.

File: lib/gateway.js
Comment (codeflyer) [reply lib/gateway.js:80]: Thanks, I'll remove it.
Yes, it needs a clean up. I'm addressing some comments and this is still in progress. 

I've reconverted the PR in `draft` to avoid it.

**Comment by codeflyer** [comment]
> * Do you have plans for TypeScript support (inc. type tests) in subsequent PRs?

Yes, the goal is to have the same stuff that are in the `mercurius` project. Probably I'll create a separate PR once every thing is stable.

> * Do you have plans for benchmarking support in subsequent PRs?

Yes.  As I wrote above, once the structure is stable and approved I'll bring in the project the same.

> * Do you have plans for more in depth docs in subsequent PRs that covers things like:
>   
>   * Plugin API Options
>   * Gateway-specific hooks
>   * Behaviour and further examples

The documentation is currently integrated in `mecurius` docs. We should decide if it worth to keep everything there or create a separated document. I'll check the docs from scratch and starting from there if there are missing examples.

> * Did you consider starting to use terminology from this proposal? https://github.com/graphql/composite-schemas-wg/pull/34/files#diff-c42a05dfff7873800ca86ea805b61b7a52aefb8f9462d0e4adb5753a037d4bfaR122 - WDYT?

I'll take a look, align the docs to this proposal it's a good idea. Maybe even some renaming in the code will help the collaborators.

> * This could also be a good opportunity to identity any existing TODOs/missing code coverage and adding related GH issues for them - WDYT?

Yes, I agree.

File: README.md
Comment (mcollina) [inline README.md:1]: update the name

File: README.md
Comment (mcollina) [inline README.md:11]: split this example in two: the gateway and the server.

File: README.md
Comment (mcollina) [inline README.md:88]: This should be named mercuriusGateway

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**
